### PR TITLE
status controller: allow setting RelatedObjects dynamically

### PR DESCRIPTION
**Pull request by squeed**
Some operators (e.g. cno) don't know the set of related objects at startup. So, allow that to be dynamically generated.

### CODE REVIEW

**Comment by squeed** [comment]
@deads2k can I get a review please? this would help us migrate CNO to library-go tooling.

File: pkg/operator/status/status_controller.go [hunk 1]
```diff
@@ -149,7 +158,16 @@ func (c StatusSyncer) Sync(ctx context.Context, syncCtx factory.SyncContext) err
 		return nil
 	}
 
-	clusterOperatorObj.Status.RelatedObjects = c.relatedObjects
+	if c.relatedObjectsFunc != nil {
+		ro := c.relatedObjectsFunc()
+		if ro != nil {
+			clusterOperatorObj.Status.RelatedObjects = ro
+		} else { // related objects unknown, carry over existing.
```
Comment (mfojtik) [inline pkg/operator/status/status_controller.go:165]: i would still append the relatedObjects from constructor? it is not clear from the NewClusterOperatorStatusController() that when WithRelatedObjectsFunc() is used the relatedObjects passed to it are dropped.

Comment (squeed) [reply pkg/operator/status/status_controller.go]: good point; updated.

**Comment by squeed** [comment]
@mfojtik updated, thanks!

**Comment by mfojtik** [comment]
/lgtm

@sttts second look/approve?

File: pkg/operator/status/status_controller.go [hunk 2]
```diff
@@ -149,7 +164,18 @@ func (c StatusSyncer) Sync(ctx context.Context, syncCtx factory.SyncContext) err
...
 
-	clusterOperatorObj.Status.RelatedObjects = c.relatedObjects
+	if c.relatedObjectsFunc != nil {
+		ro := c.relatedObjectsFunc()
+		if ro != nil {
+			// append any statically-defined related objects
+			ro = append(ro, c.relatedObjects...)
+			clusterOperatorObj.Status.RelatedObjects = ro
+		} else { // related objects (temporarily) unknown, carry over existing.
+			clusterOperatorObj.Status.RelatedObjects = originalClusterOperatorObj.Status.RelatedObjects
```
Comment (sttts) [inline pkg/operator/status/status_controller.go:174]: this has smell. Why not `c.relatedObjects`? You seem to misuse a nil return value as an error case.

File: pkg/operator/status/status_controller.go [hunk 3]
```diff
@@ -78,6 +79,20 @@ func NewClusterOperatorStatusController(
 	}
 }
 
+// WithRelatedObjectsFunc allows the set of related objects to be dynamically
+// determined.
+//
+// If the function returns nil, the set of related objects is carried over
+// from the existing ClusterOperator. This is useful if the set of related
+// objects is not yet known, perhaps because a control loop hasn't yet
+// executed. To set no related objects, return empty list.
```
Comment (sttts) [inline pkg/operator/status/status_controller.go:88]: strange signature for this behaviour. I would expect at least c.relatedObject to be applied, always.

Comment (squeed) [reply pkg/operator/status/status_controller.go:88]: Right, but then we could flap RelatedObjects.

Here's the motivating example: on restart, the CNO may wish to set a Degraded status condition immediately, but not wish to touch RelatedObjects.

Returning nil signifies "have some inertia - don't touch anything". One could, I suppose, ensure that the static `c.relatedObjects` is in the ClusterOperator object.

File: pkg/operator/status/status_controller.go (hunk 2 above)
Comment (squeed) [reply pkg/operator/status/status_controller.go:174]: It's not an error case, it's "I don't know yet". But you're right, that's kind of overloading.

I can change the signature to be more explicit.

**Comment by squeed** [comment]
@sttts updated, thanks for the review. PTAL.

File: pkg/operator/status/status_controller.go [hunk 4]
```diff
@@ -33,9 +33,12 @@ type VersionGetter interface {
 	VersionChangedChannel() <-chan struct{}
 }
 
+type RelatedObjectsFunc func() (isset bool, objs []configv1.ObjectReference)
```
Comment (sttts) [inline pkg/operator/status/status_controller.go:36]: go doc

**Pull request merged**